### ISSUE TITLE
Bug: some vertical scroll bars in code browser make screen readers freeze (fix is underway)

### ISSUE THREAD

**Issue by Palacee-hun** [NationalSecurityAgency/ghidra#2820]
There's something seriously broken with the vertical scrollbars found in Ghidra code browser tool window, but it only manifests if Ghidra is used with a screen reader or other assistive technology that queries the Java accessibility interface.
These scrollbars report that they have no parent object (AccessibleContext.indexInParent == -1), while also reporting themselves as their own parent (through AccessibleContext.getAccessibleParent). That should never happen.
When something makes these scrollbars move (e.g. some Pgdn keys are pressed in the listing panel), a value property change event is fired on them which the screen reader intercepts. It then begins to query the object for its own purposes. Among the first "questions" it asks from the Java accessibility interface is what is the top level object of the given scrollbar (should be code browser main frame). It needs this info to determine the native window handle the scrollbar belongs to. The Java-side follows the parent ancestry chain to fulfill this request, however it never terminates due to the incorrect reporting described above. A Java-side endless loop is formed wich freezes both Ghidra and the screen reader. The only way out from that is to restart the screen reader and then kill and restart the frozen Ghidra instance.
Oddly enough horizontal scrollbars in code browser window do not manifest this issue at all. They report everything correctly, and they report the scroll pane which they are in as parent (correct). I've not encountered anything like this in the main window of Ghidra, nor in any other Java app.
I've documented this issue in detail here:
https://github.com/nvaccess/nvda/issues/12045
and I also demonstrate a workaround code for NVDA screen reader that works great. However it would be best to get this fixed on Java-side. So I'd like to narrow down whether this is a Ghidra issue or a Java Swing bug (as these scrollbars are a subclass of javax.swing.JScrollBar) in order to determine where to report this issue. I know that Ghidra doesn't implement anything custom for these vertical scrollbars accessibility-wise, but I am curious what other custom info Ghidra feeds about them to Java Swing which makes that construct such mucked-up parent info about them.

**Comment by dragonmacher** [comment]
I will investigate this.    Our model for vertical scrolling in the Listing is a custom creation.

**Comment by Palacee-hun** [comment]
Thanks much. I note that not only the vertical scrollbar linked to the listing panel is affected, but e.g. the vertical scrollbar connected to the symbol tree exhibits the exact same behaviour - I diagnosed that definitely. It is that scrollbar that triggers the exact same type of freeze when expanding the functions category in the symbol tree with the right arrow key and then trying to arrow down in the functions list. And most probably other vertical scrollbars in code browser are also affected (e.g. the one linked with the data types tree - if it has one) which I haven't touched during diagnosis.

**Comment by dragonmacher** [comment]
That is helpful, thanks.

**Comment by Palacee-hun** [comment]
I can gladly help you out with that:
1. Grab a 32-bit Windows build of JRE from adoptopenjdk.net. They have it for 11.x at least (that is the version I did the diagnosis with). I recommend the zip package option.
2. In the bin folder you will find jaccessinspector-32.exe and windowsaccessbridge-32.dll. These will be your friends for debugging and reproducing the freezes. jaccessinspector is of course also part of 64-bit builds of JRE/JDK, but unfortunately that does not work, but just closes when it should display its info. That's a known bug of that, but still not fixed. That's why you need the 32-bit version. It can communicate with a 64-bit Java JDK needed for Ghidra very well. Put these two binaries (and may be jaccesswalker-32.exe, see later) into any folder of your liking.
3. Fire up Ghidra and load something into the code browser.
4. Now launch jaccessinspector. This is a simple tool that behaves very similarly to a screen reader. You can select event categories that it should monitor and upon receiving such events it will extensively query the Java accessibility API about the source object of the event and dump the findings in its window and also into a log file. For this test, tell it to monitor value changes (accessibility events menu)
5. When done you can reproduce the freezes and watch out for the info about the culprit object:
- scroll around in the listing panel with pgup/pgdn
- expand the functions category in the symbol tree and arrow down in that
6. If everything is "okay" you should get the freezes very soon: both Ghidra and jaccessinspector will stop responding. You can kill Ghidra at that time and that will reviwe jaccessinspector and you will see what info it could and couldn't get.
7. Try interacting with more scrollbars of the code browser window in various ways and compare the outputs you get from jaccessinspector for vertical and horizontal scrollbars.
8. Finally you may also find jaccesswalker-32.exe useful. That does a full query on the whole object hierarchies of all open Java windows and represents that as a tree. Do not forget to select "refresh tree" after launching it, otherwise it will present an empty tree by default (silly behaviour). You can click any object in the tree and then select "accessibility info" (or something like that) and you will get a similar dump about the object as with jaccessinspector.

**Comment by dragonmacher** [comment]
Great info!   This is a bit more involved and as such we will have to schedule time to do a more in depth examination.

**Comment by Palacee-hun** [comment]
No problem, I perfectly understand and am fully aware of the challenges this involves and thanks anyway. Diagnosing these freezes and preparing a working workaround code proposition for a screen reader was no easy task, in fact was among the tougher riddles I have ever faced in informatics and quite a lot of "hacking" was needed to get to this point.

**Comment by Palacee-hun** [comment]
I am almost sure that with very naive search approaches I have hit upon the problematic class. I am quite convinced that the vertical scrollbars that provoke havoc with screen readers are the instances of docking.widgets.SideKickVerticalScrollbar. This class has an (at least to me) quite confusing delegation pattern and hard-to-follow parent relationships. Furthermore it has the following code fragment:
```
	@Override
	public AccessibleContext getAccessibleContext() {
		return delegate.getAccessibleContext();
	}

```
The whole Ghidra code tree has only three mentions of AccessibleContext, and this fragment is eventually found in a class that implements a type of vertical scrollbar (the two other mentions are in totally unrelated places). I don't see the point of this code fragment as this scrollbar is a subclass of jScrollbar, so handling of accessibility is already transparent as with subclasses of other regular Swing controls (like combo boxes or push buttons) for which Ghidra doesn't do anything similar (and they do work all right). What's more this fragment is indeed a possible source of an endless recursion.
So I suggest focusing investigation efforts to this class, maybe together with the package docking.widgets.indexedscrollpane, the only place where this sidekick vertical scrollbar thing is incorporated.

**Comment by dragonmacher** [comment]
Thanks for the tip.  I started by looking there and at our GTree implementation, which is a panel that contains an `AutoScrollTree`.  I'm assuming that there is something wrong with the auto scroll tree too.    Is the problem for vertical scroll bars for many of Ghidra's trees?

**Comment by Palacee-hun** [comment]
I don't know how many of the tree controls are affected. What I do know from the diagnosis period is that the culprit vertical scroll bar is both present near the listing panel and also the symbol tree and is a subclass of jScrollbar with some custom features and it also resides in a scroll pane. After having looked at its source code I see that this new "surprise finding" of this sidekick vertical scroll bar class fulfills all these criteria.

**Comment by Palacee-hun** [comment]
My instincts and reasoning has proven to be right when I've put it to the test. I hand-picked SideKickVerticalScrollbar.java out of docking-src.zip found in the Ghidra release. I removed just the method cited above as I suspected that it was the only cause of all these freezes. I recompiled and inserted the resulting three class files back into docking.jar found in the release overwriting the original three class files there. Hackish, isn't it? Yes definitely, but it's the fastest way for such a small fix by a large margin.
Then I disengaged my workaround code from the screen reader, so it should have frozen up like before if I hadn't been on the right track. But it didn't. No matter how hard I'd been trying to reproduce the crashes, it was in vain, all of them were banished by this small fix and the code browser worked normally. I did all my diagnostics again (among them Jaccessinspector described above) and the problematic vertical scrollbars turned into absolutely normal scrollbars and spat out correct accessibility information. No more reporting of self as parent, parent and top level object is valid and correct, all is well.
So that was the whole story: one little method (most probably some left-over code) with 
faulty delegation pattern that lead the accessibility interface astray. And this method is totally pointless as javax.swing.JScrollbar takes care of it on its own. So it can be safely removed from Ghidra and this tough issue is gone. I experience no "side-effects" from the removal and there shouldn't be any as the code is pointless.

**Comment by Palacee-hun** [comment]
As conclusion and for documentation I paste here the output of Jaccessinspector upon a value change on the sidekick vertical scrollbar attached to the listing panel.
First the output when querying the scroll bar produces endless loop and freeze (see comment):
```
Value changed event: old = 0; new = 850226

Version Information:
    Java virtual machine version: 11.0.7
    Access Bridge Java class version: 11.0.7
    Access Bridge Java DLL version: 11.0.7
    Access Bridge Windows DLL version: 11.0.7

AccessibleContext information:
    Name:  
    Virtual Name:  
    Description:  
    Role:  scroll bar
    Role in en_US locale:  scroll bar
    States:  enabled,focusable,visible,showing,opaque,vertical
    States in en_US locale:  enabled,focusable,visible,showing,opaque,vertical
    Index in parent:  -1
    Children count:  2
    Bounding rectangle:  [530, 159, 547, 729]
ERROR: GetAccessibleContextInfo failed for top-level window // freeze takes place here, output is produced after killing Ghidra, further lines are meaningless
    No parent
ERROR: GetVisibleChildrenCount failed
GetAccessibleRelationSet failed.

Accessible Value information:
```
And now the correct info about the same scrollbar after removing the little faulty method by hand:
```
Value changed event: old = 447844; new = 446165

Version Information:

    Java virtual machine version: 11.0.7
    Access Bridge Java class version: 11.0.7
    Access Bridge Java DLL version: 11.0.7
    Access Bridge Windows DLL version: 11.0.7

AccessibleContext information:

    Name:  
    Virtual Name:  
    Description:  
    Role:  scroll bar
    Role in en_US locale:  scroll bar
    States:  enabled,focusable,visible,showing,opaque,vertical
    States in en_US locale:  enabled,focusable,visible,showing,opaque,vertical
    Index in parent:  0
    Children count:  2
    Bounding rectangle:  [530, 159, 547, 729]
    Top-level window name:  CodeBrowser: procwin:/Procwin.EXE
    Top-level window role:  frame
    Parent name:  
    Parent virtual name:  
    Parent role:  scroll bar
    Visible descendents count:  2
    Descendent 0 name:  
    Descendent 0 role:  push button
    Descendent 1 name:  
    Descendent 1 role:  push button

Accessible Value information:
    Current Value:  446165
    Maximum Value:  1398810
    Minimum Value:  0
```

**Comment by dragonmacher** [comment]
Nice work.  

> I note that not only the vertical scrollbar linked to the listing panel is affected, but e.g. the vertical scrollbar connected to the symbol tree exhibits the exact same behaviour

The symbol tree does not use the `SideKickVerticalScrollbar', AFAIK.    Were you able to verify if the tree is still broken in your tests?

**Comment by Palacee-hun** [comment]
Of course, the symbol tree is okay too and this little fix stops all freeze scenarios we know of (it was not just me who did the reproduction tests at the beginning of diagnosis).
Yes, as I opened my Jaccessinspector logs again, I see that the vertical scrollbar near the symbol tree that used to trigger freezes doesn't report exactly in the same manner as sidekick vertical scrollbars do. Jaccessinspector data has some strangeness for that, the most particular is that it reports [-1; -1; -2; -2] as bounding rectangle coordinates. But the screen reader doesn't panic on that, and figures out the correct values (it has more logic in it than Jaccessinspector and is usually extremely fault-tolerant). The important point is that it doesn't exhibit any self-parenting behaviour any more, my fix cured that also somehow. The screen reader only detects that vertical scrollbar if I expand e.g. the functions category in symbol tree, if it is closed, then only a horizontal scrollbar is presented.
